### DISCUSSION TITLE
How to get the translated slugs on the front-end

### DISCUSSION THREAD

**Discussion by mcamprecios** [qtranslate/qtranslate-xt#1401] [Q&A]
Hi!

I'm trying to update an old project that used qTranslate-X and qTranslate Slugs, but i'm getting lost... 

Basically, there are many calls to the ```qts_get_slug()``` function, and I don't know how to replace it. My editor says that ```get_slug()``` is available and I fould it as a public function. But I'm not sure how to call it.

It's mostly shown as ```qts_get_slug($r->ID, $currentLang)```. 

I replaced the previous plugins with the new one, deactivated everything else related to the old plugin and migrated the slugs.

What am I missing? I'm sorry, I'm not a PHP dev.

Thanks a lot for your work!

**Comment by mcamprecios** [comment]
Just one more comment, because I may be messing around... 

I do know that activating "qTranslate Slugs" makes it work... But with some warnings. I'm already using PHP 7.4 insteaad of 8.1 or 8.2. 

On the admin:

```The plugin generated 299 characters of unexpected output during activation. If you notice “headers already sent” messages, problems with syndication feeds or other issues, try deactivating or removing this plugin.```

On the front:

``` Deprecated: Methods with the same name as their class will not be constructors in a future version of PHP; QtranslateSlugWidget has a deprecated constructor in D:\Repos\lantana\wp-content\plugins\qtranslate-slug\includes\class-qtranslate-slug-widget.php on line 8```

Sounds weird to me because my understanding is that "qTranslte Slugs" was ported into qTranslate-xt, instead of creating a bridge.

Thanks again!

**Comment by spleen1981** [comment]
Migration should work with this sequence:

- [make a backup of the database]
- make sure you are on latest version of legacy plugin
- deactivate the legacy plugin
- migrate using `Migrate QTS slugs` in QTX settings>import/export
- activate slugs module
- delete the legacy plugin

The above only migrates the legacy plugin options, if in your project there are calls to the legacy plugin functions, those need to be replaced manually, based on current slugs module source and your project intents.
E.g. `qts_get_slug` function is probably to be replaced with `get_slug` function in the `QTX_Module_Slugs` class, if that is the case you can use it as follows:

```
global $qtranslate_slugs;
$slug = $qtranslate_slugs->get_slug( $id, $lang );
```

**Comment by mcamprecios** [comment]
Hi @spleen1981 

Thanks a lot for the update, it did work like a charm besides a couple of minor issues.

I've seen that you are one of the module's developer. I believe it may be a good idea to add your instructions to the [Slugs readme file](https://github.com/qtranslate/qtranslate-xt/tree/master/src/modules/slugs). 

Should I do a pull request?

**Reply by spleen1981** [reply]
Sure, you can submit a PR adding the bullets above to detail migration process.
No need to describe the use of `QTX_Module_Slugs` methods instead, that is already documented in the class definition.

**Comment by mcamprecios** [comment]
Thanks @spleen1981 

I'm wondering if the order matters and it is only possible to **migrate the slugs after activating the module**. Otherwise the option doesn't show up in *QTX Settings > Import/Export*. I may be wrong, our site is quite dated...

If it is true, and given that I never used the original plugin and I'm mostrly unfamiliar with it all, I would replace the actual "I've been using the QTS plugin, how can I find the existing slugs?" section with:

### How can I migrate from the legacy QTS plugin to the new module?

If you are migrating from using *qTranslate X* and the *qtranslate-slug* (QTS) plugin, migration should work with this sequence:

1. Make sure that you are on **latest version of the legacy plugin**
2. **Deactivate the legacy plugin**
3. **Activate the *Slugs* module**
4. **Migrate** using `Migrate QTS slugs` in *QTX Settings > Import/Export*. The dry-run mode allows to test the import and see the number of rows changed before updating the database.
5. **Delete the legacy plugin**

If you are using functions from the legacy plugin in your theme files, you may want to switch to a neutral theme.

**Attention!** Do not uninstall QTS before migrating data to qTranslate, it deletes data permanently. Once the migration is done, the data is not be visible in QTS anymore and the QTS plugin can be uninstalled.

**Comment by spleen1981** [comment]
> I'm wondering if the order matters and it is only possible to migrate the slugs after activating the module.

Yes the slugs module must be activated to migrate, as relevant code is part of the module itself. Hence activation bullet come first, and migration later.

**Reply by mcamprecios** [reply]
Thanks a lot, pull request done.

**Comment by herrvigg** [comment]
Merged PR #1404 👍
